storage/api.go

```diff
@@ -7,11 +7,17 @@ import (
 
 // --- basics --->
 
+type Storage interface {
+	Has(ctx context.Context, key string) (bool, error)
```

Comment:
recap:  The holistic design goal I'm keeping aim on is: to consistently tell both users and intermediate API designers that they should be referring to either `ReadableStorage` or `WritableStorage`.  And nothing else -- because then they should use the package-scope functions -- *not* the methods -- in order to do all their work (and let those functions take care of all the feature detection for them).

Now, `Has` is an operation that a user might want to ask for on either one of them.

(And surprisingly, I'd say one is almost more likely to want to ask `Has` on the writable side.  To ask `Has` on the reader might be slightly cheaper than opening the read stream for some implementations; but to ask `Has` before _writing_ can be _significantly_ cheaper than opening a write stream for some implementations.  But I digress; let's suffice to say we want the operation available on either direction.)

Now if I want to make a package-scope function for `Has`, just to keep 100% consistent with the messaging to the API user about "use the package-scope functions for all operations"... it turns out to be the one thing so far that's legitimately in common to both the read and the write directions.

If golang had parametric polymorphism, I'd write two functions:

```
func Has(Context, ReadableStorage, Key) (bool, error) {...}
func Has(Context, WritableStorage, Key) (bool, error) {...}
```

... but we do not have parametric polymorphism, so this is not an option.

Then our other options are:

- Give up on having a package-scope `Has` function, and document the inconsistency of the "use the package-scope functions for all operations" rule for users.
- Probably still have `Has` as _methods_ on both anyway?
- Or if not, and just putting `Has` on `ReadableStorage` only, then... we make people feature-detect `Has` on the writer side, when that's wanted?
- Or we throw in the towel on separating reader and writer directions entirely?  But I don't care for this at all -- this split has felt really good so far.  It's like a soft form of "capabilities"-style APIs that keeps you from making mistakes in wiring, and it's also really great to be able to implement a read-only system and express that clearly at compile time.  So surely we don't want to give all that up.

All of those choices sound worse to me than introducing an interface that's at the bottom of the type graph here.

Introducing an interface type that's at the bottom of the type graph gives us what we need to make the package API consistent and low-friction.  Coincidentally, `Has` fits really nicely there.  (And in the future: if we find more features that might work on both the read or the write direction (and presumably, are optional extensions), we make those package-scope functions of the style `func Foo(Context, Storage, ...) (...)` too, reusing the same bottom type and doing feature-detection internally from there, to minimize burden to the caller.)

Reply:
(This is sort of like the inverse of https://github.com/ipld/go-ipld-prime/pull/277 .  It makes the most sense if you look at how many casts show up when you actually try to _use_ the API.)